### PR TITLE
Add `ASYNCIFY_PROPAGATE_ADD` setting

### LINKED ISSUE #13150: Make ASYNCIFY_ADD propagate (or add a setting to do so)
An app I'm porting has 3000 instrumented functions when ASYNCIFY_IGNORE_INDIRECT is disabled, and 100 instrumented functions when it is enabled. Somewhere in between is the actual number of functions that need to be instrumented.

After investigating the code I found one function with indirect calls that does need to be instrumented (there will probably be more), so I specified it in ASYNCIFY_ADD. But then I realised that its status is not propagated upwards, I would need to specify all the functions that call it, etc. I tried adding it to ASYNCIFY_IMPORTS instead, but it was ignored because it's not an import function.

So firstly, have I understood this correctly, that ASYNCIFY_ADD doesn't propagate? (I think ASYNCIFY_REMOVE actually does.) Second, what is the reason for this? Can ASYNCIFY_ADD be changed to propagate? Or could a new setting which propagates be added? I think that would be much more useful and productive than specifying each function in the call tree.

### PR DESCRIPTION
**Pull request by curiousdannii**
Adds a `ASYNCIFY_PROPAGATE_ADD` setting to control whether the `ASYNCIFY_ADD` list propagates. Closes #13150

I set this to be on by default, which does change the behaviour of `ASYNCIFY_ADD`. I think having it off by default is unsafe, and also greatly reduces its usefulness. For example, if you worked out the add-list in release mode, then switched back to debug mode, then it's likely your app would stop working because previously inlined functions would now need to be instrumented.

There are few examples in public code of people actually using it, probably because it requires you to list everything above it in the call tree. Here's one of the few examples I could find, which could probably be substantially simplified by this: https://github.com/Siv3D/OpenSiv3D/blob/c81400c5a39405d66843c07ce1cf90f46e01a86a/Web/App/CMakeLists.txt#L29
Thanks to @nokotan for getting the code working in Binaryen!

Note: I don't know if the Binaryen changes are in Emscripten yet.

### CODE REVIEW

File: ChangeLog.md
Comment (kripken) [inline ChangeLog.md:47]:
```suggestion
  behaviour, disable that setting. (#21672)
```

**Review by kripken — COMMENTED** [review]
Thanks! In general this looks good, though I still lean towards having this off by default. However, I don't feel strongly, so if most people want it on that sounds fine. Maybe let's wait a few days to see if people have opinions here.

If we do want it on then perhaps the changelog entry could be enlarged a little to say something like "as a result of this change, you may see an increase in the size of ASYNCIFY builds if you use ASYNCIFY_ADD (but correctness is not affected). To undo that and return to the old behavior, set `-sNO_ASYNCIFY_PROPAGATE_ADD`." (that is, to concretely explain what users might see changes)

**Comment by curiousdannii** [comment]
> as a result of this change, you may see an increase in the size of ASYNCIFY builds if you use ASYNCIFY_ADD (but correctness is not affected)

I was trying to think of how it could end up bigger, because previously you would have to have still listed the functions above in ASYNCIFY_ADD. But suppose you have these functions:

```
void a() { c(); }
void b() { c(); }
void c() {  }
```

If you put `c` in ASYNCIFY_ADD, then both `a` and `b` will be added. But you might know that while `a` results in an async import being called, `b`'s use of `c` never does. But the compiler can't work that out, so it has to add `b`. Previously you had to add `a` manually to be safe. Now you can manually add `b` to ASYNCIFY_REMOVE. So propagation by default is safer, but can result in bigger builds. Optimal use may still need some functions to be added (to REMOVE instead of ADD), but this is much safer, as, for example, different inlining decisions leading to a function not being instrumented could be fatal, but inlining decisions leading to a function being instrumented won't be.

I'll edit the changelog to explain this a little better.

**Comment by kripken** [comment]
Changelog looks good to me, thanks.

The CI error suggests to me that you need to run `tools/maint/update_settings_docs.py` to update a file, but I'm not sure.

**Comment by curiousdannii** [comment]
I don't like checked in autogenerated docs... I edited it manually haha.

**Comment by kripken** [comment]
That does look like a random test error, I restarted it.

Let's give it another day or so to see if people have thoughts on the default here. Otherwise I'm ok to land it as is, my preference the other way is minor.

**Comment by curiousdannii** [comment]
Has the changes from Binaryen been imported here yet? It wouldn't make sense to merge this until that happens..

**Comment by kripken** [comment]
Ah, yes, we need to wait for that, and while I pressed auto-merge it looks like that hasn't finished yet. But it should finish soon and the update then reaches `emsdk tot` in a matter of an hour or so.

**Comment by kripken** [comment]
I realized it's simple to test this, so I suggest we add that in this PR before landing. This should work, I think:
```diff
diff --git a/test/test_core.py b/test/test_core.py
index a6b437a0b..5f96694b9 100644
--- a/test/test_core.py
+++ b/test/test_core.py
@@ -8215,7 +8215,10 @@ Module.onRuntimeInitialized = () => {
   @parameterized({
     'normal': ([], True),
     'ignoreindirect': (['-sASYNCIFY_IGNORE_INDIRECT'], False),
-    'add': (['-sASYNCIFY_IGNORE_INDIRECT', '-sASYNCIFY_ADD=["__original_main","main","virt()"]'], True),
+    'add': (['-sASYNCIFY_IGNORE_INDIRECT', '-sASYNCIFY_ADD=["virt()"]'], True),
+    # If ASYNCIFY_PROPAGATE_ADD is disabled then we must specify the callers of
+    # virt() manually, rather than have them inferred automatically.
+    'add_no_prop': (['-sASYNCIFY_IGNORE_INDIRECT', '-sASYNCIFY_ADD=["__original_main","main","virt()"]', '-sASYNCIFY_PROPAGATE_ADD=0'], True),
   })
   def test_asyncify_indirect_lists(self, args, should_pass):
     self.set_setting('ASYNCIFY')
```
The Binaryen change has [rolled in](https://chromium-review.googlesource.com/c/emscripten-releases/+/5435643), but it takes a bit for new builds of tip of tree to be uploaded. Probably in an hour or two from now that should work.

**Review by kripken — APPROVED** [review]

**Comment by sbc100** [comment]
It looks like this change causes the`lto2.test_asyncify_indirect_lists_add` test to fail.

**Comment by kripken** [comment]
I think we just need to disable that test in LTO mode, I opened https://github.com/emscripten-core/emscripten/pull/21765 now.

**Comment by curiousdannii** [comment]
So months after adding propagating ASYNCIFY_ADD, I'm finally trying it myself. My understanding was that you could enable `ASYNCIFY_IGNORE_INDIRECT` as long as you ensured that whenever an actual async function is run, that all functions in the tree above it are also instrumented. But now that I'm trying it, I'm still getting errors. I'm not sure it's really feasible to disable `ASYNCIFY_IGNORE_INDIRECT` and then to ADD specific functions. This work may have been for nothing 😞

**Pull request merged**
